torch/lib/THD/base/tensors/generic/THTensor.cpp

```diff
@@ -209,8 +215,268 @@ auto THTensor<real>::free() -> THTensor& {
+  template<>                                                         \
+  auto THTensor<real>::NAME##ValueT(const Tensor& t,                 \
+                                   scalar_type value) -> THTensor& { \
+    THTensor_(NAME##ValueT)(                                         \
+      tensor,                                                        \
+      non_const_cast(t).tensor,                                      \
+      value                                                          \
+    );                                                               \
+    return *this;                                                    \
+  }                                                                  \
+                                                                     \
+  template<>                                                         \
+  auto THTensor<real>::NAME##Tensor(const Tensor& r,                 \
+                                    const Tensor& tb) -> THTensor& { \
+    if (r.type() != Type::UCHAR)                                     \
+      throw std::invalid_argument("logical operator called on non-byte tensor"); \
+    THTensor_(NAME##Tensor)(                                         \
+      non_const_byte_cast(r).tensor,                                 \
+      tensor,                                                        \
+      non_const_cast(tb).tensor                                      \
+    );                                                               \
+    return *this;                                                    \
+  }                                                                  \
+                                                                     \
+  template<>                                                         \
+  auto THTensor<real>::NAME##TensorT(const Tensor& ta,               \
+                                     const Tensor& tb) -> THTensor& { \
+    THTensor_(NAME##TensorT)(                                        \
+      tensor,                                                        \
+      non_const_cast(ta).tensor,                                     \
+      non_const_cast(tb).tensor                                      \
+    );                                                               \
+    return *this;                                                    \
+  }                                                                  \
+
+TENSOR_IMPLEMENT_LOGICAL(lt)
+TENSOR_IMPLEMENT_LOGICAL(gt)
+TENSOR_IMPLEMENT_LOGICAL(le)
+TENSOR_IMPLEMENT_LOGICAL(ge)
+TENSOR_IMPLEMENT_LOGICAL(eq)
+TENSOR_IMPLEMENT_LOGICAL(ne)
+
+#undef TENSOR_IMPLEMENT_LOGICAL
+
+template<>
+auto THTensor<real>::abs(const Tensor& src) -> THTensor& {
+#if defined(TH_REAL_IS_INT) || defined(TH_REAL_IS_LONG)
+  THTensor_(abs)(tensor, non_const_cast(src).tensor);
+  return *this;
+#else
+  throw std::runtime_error("absolute value is only available for `int` and `long`");
```

Comment:
Yeah, I think it's defined twice, but within exclusive `ifdef`s. That's where the confusion comes from.